warp10/src/main/java/io/warp10/script/functions/MOTIONSPLIT.java

```diff
@@ -32,174 +33,303 @@
     double proximityZoneTraveledDistance = 0.0D;
-    
+
     while (idx < n) {
       long timestamp = GTSHelper.tickAtIndex(gts, idx);
       long location = GTSHelper.locationAtIndex(gts, idx);
       long elevation = GTSHelper.elevationAtIndex(gts, idx);
       Object value = GTSHelper.valueAtIndex(gts, idx);
-    
+
+      //
+      // Initialise refLocation as soon as there is a valid position
+      //
+      if (GeoTimeSerie.NO_LOCATION == refLocation && GeoTimeSerie.NO_LOCATION != location) {
+        refLocation = location;
+        refTimestamp = timestamp;
+      }
+
+      //
+      // If the previous tick was more than 'timeThreshold' ago, split now
+      //
+      if (timestamp - previousTimestamp > timeThreshold) {
+        if (null != splitTypeLabel) {
+          split.getMetadata().putToLabels(splitTypeLabel, SPLIT_TYPE_TIME);
+        }
+        mustSplit = true;
+      }
+
       //
-      // If the previous tick was more than 'timeThreashold' ago, split now
       // If the distance to the previous location is above distanceThreshold, split
       //
-      
-      if (timestamp - previousTimestamp > timeThreshold
-          || (GeoTimeSerie.NO_LOCATION != previousLocation && GeoTimeSerie.NO_LOCATION != location && GeoXPLib.orthodromicDistance(location, previousLocation) > distanceThreshold)) {
+      if (GeoTimeSerie.NO_LOCATION != previousLocation && GeoTimeSerie.NO_LOCATION != location && GeoXPLib.orthodromicDistance(location, previousLocation) > distanceThreshold) {
+        if (null != splitTypeLabel) {
+          split.getMetadata().putToLabels(splitTypeLabel, SPLIT_TYPE_DISTANCE);
+        }
         mustSplit = true;
       }
-      
+
       //
-      // If the current point is farther away from the reference point than 'proximityZoneRadius'
+      // If the current point is farther away from the reference point than 'proximityZoneRadius', or the instant speed is greater than proximityZoneSpeed
       // change the reference point and reset the traveled distance.
-      // If the distance traveled in the proximity zone was traveled at less than proximityZoneSpeed and the time spent in the zone is above proximityZoneTime, force a split
+      // If the distance traveled in the proximity zone was traveled at less than proximityInZoneMaxMeanSpeed and the time spent in the zone is above proximityZoneTime, force a split
       //
-      
-      if (GeoTimeSerie.NO_LOCATION != refLocation && GeoTimeSerie.NO_LOCATION != location && GeoXPLib.orthodromicDistance(refLocation, location) > proximityZoneRadius) {
-        if (previousTimestamp - refTimestamp > proximityZoneTime && proximityZoneTraveledDistance / ((double) (previousTimestamp - refTimestamp) / Constants.TIME_UNITS_PER_S) < proximityZoneSpeed) {
-          mustSplit = true;
+      long timeStopped = previousTimestamp - refTimestamp;
+      if (GeoTimeSerie.NO_LOCATION != refLocation && GeoTimeSerie.NO_LOCATION != location && null != split) {
```

Comment:
Why checking for nullity of split here and not for the other tests?

Reply:
Because I write the stopped time label later. It avoids an exception on a specific case:

```
NEWGTS
1 1 1 1 T ADDVALUE
{
  'stopped.min.time'  0 s
  'stopped.max.speed' 0 
  'stopped.max.radius' -1
  'label.stopped.time' 'stop_time'
  'label.split.start' 'split'
  'label.split.type' 'reason'
}
MOTIONSPLIT
```